File: webflux/webclient/aggregate-response/src/main/java/br/com/leonardoferreira/webflux/service/AggregateService.java
package br.com.leonardoferreira.webflux.service;

import br.com.leonardoferreira.webflux.domain.AggregatedResponse;
import br.com.leonardoferreira.webflux.domain.AnythingRequest;
import br.com.leonardoferreira.webflux.domain.AnythingResponse;
import br.com.leonardoferreira.webflux.domain.TodoResponse;
import br.com.leonardoferreira.webflux.domain.UuidResponse;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

@Service
public class AggregateService {

    private final WebClient httpBinClient;

    private final WebClient jsonPlaceHolderClient;

    public AggregateService(final WebClient httpBinClient,
                            final WebClient jsonPlaceHolderClient) {
        this.httpBinClient = httpBinClient;
        this.jsonPlaceHolderClient = jsonPlaceHolderClient;
    }

    public Mono<AggregatedResponse> call() {
        final Mono<TodoResponse> todoResponseMono = httpBinClient.post()
                .uri("/anything")
                .body(Mono.just(new AnythingRequest(200)), AnythingRequest.class)
                .retrieve()
                .bodyToMono(AnythingResponse.class)
                .map(AnythingResponse::getJson)
                .map(AnythingResponse.Json::getNumber)
                .flatMap(this::retrieveTodo);

        final Mono<UuidResponse> uuidResponseMono = httpBinClient.get()
                .uri("/uuid")
                .retrieve()
                .bodyToMono(UuidResponse.class);

        return Mono.zip(todoResponseMono, uuidResponseMono, this::buildAggregatedResponse);
    }

    private AggregatedResponse buildAggregatedResponse(final TodoResponse todoResponse, final UuidResponse uuidResponse) {
        return new AggregatedResponse(
                todoResponse.getTitle(),
                uuidResponse.getUuid()
        );
    }

    private Mono<TodoResponse> retrieveTodo(final Integer number) {
        return jsonPlaceHolderClient.get()
                .uri(uriBuilder -> uriBuilder.path("/todos/{number}").build(number))
                .retrieve()
                .bodyToMono(TodoResponse.class);
    }


}

File: webclient/with-kotlin/src/main/kotlin/br/com/leonardoferreira/poc/service/AggregateService.kt
package br.com.leonardoferreira.poc.service

import br.com.leonardoferreira.poc.domain.request.AnythingRequest
import br.com.leonardoferreira.poc.domain.response.AggregatedResponse
import br.com.leonardoferreira.poc.domain.response.AnythingResponse
import br.com.leonardoferreira.poc.domain.response.TodoResponse
import br.com.leonardoferreira.poc.domain.response.UuidResponse
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.runBlocking
import org.springframework.stereotype.Service
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.awaitBody
import org.springframework.web.reactive.function.client.awaitExchange

@Service
class AggregateService(
    private val httpBinClient: WebClient,
    private val jsonPlaceHolderClient: WebClient
) {

    suspend fun call(): AggregatedResponse = coroutineScope {
        val todoAsync = async { retrieveTodo() }

        val uuidResponse = async { retrieveUuid() }

        AggregatedResponse(
            todoTitle = todoAsync.await().title,
            uuid = uuidResponse.await().uuid
        )
    }

    private suspend fun retrieveUuid(): UuidResponse =
        httpBinClient.get()
            .uri("/uuid")
            .awaitExchange()
            .awaitBody()

    private suspend fun retrieveTodo(): TodoResponse {
        val anythingResponse = httpBinClient.post()
            .uri("/anything")
            .bodyValue(AnythingRequest(200))
            .awaitExchange()
            .awaitBody<AnythingResponse>()

        return jsonPlaceHolderClient.get()
            .uri { uriBuilder ->
                uriBuilder.path("/todos/{id}")
                    .build(anythingResponse.json.number)
            }
            .awaitExchange()
            .awaitBody()
    }

}
